File: Leetcode/Java Codes/October/_16.java
import java.util.*;

class _16 {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] nums = new int[n];
        for(int i=0; i<n; i++){
            nums[i] = sc.nextInt();
        }
        int value = sc.nextInt();
        System.out.println(helper(nums,value));
        sc.close();
    }

    static int helper(int[] nums, int value){
        HashMap<Integer, Integer> remainders = new HashMap<>();

        for(int num=0; num<nums.length; num++){
            int remainder = ((nums[num]%value) + value) % value;

            remainders.put(remainder, remainders.getOrDefault(remainder,0)+1);
        }

        int x=0;
        while(true){

            int rem = x%value;
            if(!remainders.containsKey(rem)){
                break;
            } else {
                int count = remainders.get(rem);
                if(count == 1){
                    remainders.remove(rem);
                } else {
                    remainders.put(rem, count - 1);
                }
                x++;
            }
        }

        return x;
    }
}

File: Leetcode/Python/_16.py
class Solution:
    def helper(self,nums, value) -> int:
        n = len(nums)
        remainders = {}

        for i in range(n):
            rem = ((nums[i]%value)+value)%value
            remainders[rem] = remainders.get(rem,0)+1

        x=0
        while True:
            rem = x%value
            if rem not in remainders :
                break
            else :
                count = remainders.get(rem)
                if count == 1:
                    remainders.pop(rem)
                else :
                    remainders[rem] = count - 1
             
                x += 1

        return x
    
def main():
    nums = list(map(int, input().split()))
    value = int(input())
    print(Solution().helper(nums,value))

if __name__ == "__main__":
    main()